File: java/11CompleteImplementationCode/src/main/java/com/neomatrix369/sorter/SortByFrequencyDescAndByWordAlphabetically.java
package com.neomatrix369.sorter;

import com.neomatrix369.domain.WordEntity;

import java.util.Comparator;

public class SortByFrequencyDescAndByWordAlphabetically implements Comparator<WordEntity> {

    private static final int THAT_THE_FIRST_WORD_HAS_LOWER_FREQUENCY = 1;
    private static final int THAT_THE_FIRST_WORD_HAS_HIGHER_FREQUENCY = -1;

    @Override
    public int compare(WordEntity firstWordEntity, WordEntity secondWordEntity) {
        // Sort Frequency in DESC order
        if (firstTheWordIsLessFrequentlyRepeatedThanTheSecond(
                firstWordEntity, secondWordEntity))
            return THAT_THE_FIRST_WORD_HAS_LOWER_FREQUENCY; // greater than

        if (firstTheWordIsMoreFrequentlyRepeatedThanTheSecond(
                firstWordEntity, secondWordEntity))
            return THAT_THE_FIRST_WORD_HAS_HIGHER_FREQUENCY; // less than

        // equal to
        // same frequency
        // sort alphabetically
        return frequenciesAreSameLetsSortThemByWordsAlphabetically(
                firstWordEntity, secondWordEntity);
    }

    private int frequenciesAreSameLetsSortThemByWordsAlphabetically(
            WordEntity firstWordEntity, WordEntity secondWordEntity) {
        return firstWordEntity.getWord()
                .compareTo(secondWordEntity.getWord());
    }

    private boolean firstTheWordIsMoreFrequentlyRepeatedThanTheSecond(
            WordEntity firstWordEntity, WordEntity secondWordEntity) {
        if (firstWordEntity.getFrequency() >
                secondWordEntity.getFrequency()) {
            return true;
        }
        return false;
    }

    private boolean firstTheWordIsLessFrequentlyRepeatedThanTheSecond(
            WordEntity firstWordEntity, WordEntity secondWordEntity) {
        if (firstWordEntity.getFrequency() <
                secondWordEntity.getFrequency()) {
            return true;
        }
        return false;
    }
}

File: csharp/11CompleteImplementationCode/SortByFrequencyDescAndByWordAlphabetically.cs
using System;
using System.Collections.Generic;

namespace CompleteImplementationCode
{
	public class SortByFrequencyDescAndByWordAlphabetically : IComparer<WordEntity> {

		private const int THAT_THE_FIRST_WORD_HAS_LOWER_FREQUENCY = 1;
		private const int THAT_THE_FIRST_WORD_HAS_HIGHER_FREQUENCY = -1;

		public int Compare(WordEntity firstWordEntity, WordEntity secondWordEntity) {
			// Sort Frequency in DESC order
			if (FirstTheWordIsLessFrequentlyRepeatedThanTheSecond(
				firstWordEntity, secondWordEntity))
				return THAT_THE_FIRST_WORD_HAS_LOWER_FREQUENCY; // greater than

			if (FirstTheWordIsMoreFrequentlyRepeatedThanTheSecond(
				firstWordEntity, secondWordEntity))
				return THAT_THE_FIRST_WORD_HAS_HIGHER_FREQUENCY; // less than

			// equal to
			// same frequency
			// sort alphabetically
			return FrequenciesAreSameLetsSortThemByWordsAlphabetically(
				firstWordEntity, secondWordEntity);
		}

		private int FrequenciesAreSameLetsSortThemByWordsAlphabetically(
			WordEntity firstWordEntity, WordEntity secondWordEntity) {
			return firstWordEntity.Word
				.CompareTo(secondWordEntity.Word);
		}

		private bool FirstTheWordIsMoreFrequentlyRepeatedThanTheSecond(
			WordEntity firstWordEntity, WordEntity secondWordEntity) {
			if (firstWordEntity.Frequency >
				secondWordEntity.Frequency) {
				return true;
			}
			return false;
		}

		private bool FirstTheWordIsLessFrequentlyRepeatedThanTheSecond(
			WordEntity firstWordEntity, WordEntity secondWordEntity) {
			if (firstWordEntity.Frequency <
				secondWordEntity.Frequency) {
				return true;
			}
			return false;
		}
	}
}

